File: app/src/main/java/kagura/project/com/a8/association/memorielle/MemoryGame.java
package kagura.project.com.a8.association.memorielle;


import android.animation.AnimatorInflater;
import android.animation.AnimatorSet;
import android.annotation.SuppressLint;
import android.content.DialogInterface;
import android.content.SharedPreferences;
import android.content.pm.ActivityInfo;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.os.SystemClock;
import android.preference.PreferenceManager;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.AdapterView;
import android.widget.Chronometer;
import android.widget.GridView;
import android.widget.ImageView;
import android.widget.RelativeLayout;

import com.plattysoft.leonids.ParticleSystem;

import java.util.Calendar;
import java.util.List;
import java.util.Timer;
import java.util.TimerTask;

import kagura.project.com.a8.association.Association;
import kagura.project.com.a8.adapters.ImageAdapter;
import kagura.project.com.a8.R;
import kagura.project.com.a8.association.ResultFragment;
import kagura.project.com.a8.database.ResultDAO;
import kagura.project.com.a8.collections.Card;
import kagura.project.com.a8.collections.Result;

public class MemoryGame extends AppCompatActivity {

    private static int level;

    private boolean isClickable = true;

    private int finish;
    private int tries = 0;

    private UpdateCardsHandler handler;
    private Card firstCard, secondCard;
    private AnimatorSet setRightOutFirst, setRightInFirst, setRightOutSecond, setRightInSecond;
    private static final Object lock = new Object();

    GridView gridviewFront, gridviewBack;
    Fragment fragmentResult;

    ImageView buttonBack;
    RelativeLayout relativeLayout;

    Boolean isTimerStarted = false;
    Chronometer timer;
    Calendar calendar;

    FragmentManager fragmentManager;
    Association memory;


    public MemoryGame() {
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {

        super.onCreate(savedInstanceState);


        requestWindowFeature(Window.FEATURE_NO_TITLE);
        getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN,
                WindowManager.LayoutParams.FLAG_FULLSCREEN);

        getWindow().requestFeature(Window.FEATURE_ACTION_BAR);
        //noinspection ConstantConditions
        getSupportActionBar().hide();
        setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE);

        setRightOutFirst = (AnimatorSet) AnimatorInflater.loadAnimator(getApplicationContext(),
                R.animator.card_flip_right_out);

        setRightInFirst = (AnimatorSet) AnimatorInflater.loadAnimator(getApplicationContext(),
                R.animator.card_flip_right_in);

        setRightOutSecond = (AnimatorSet) AnimatorInflater.loadAnimator(getApplicationContext(),
                R.animator.card_flip_right_out);

        setRightInSecond = (AnimatorSet) AnimatorInflater.loadAnimator(getApplicationContext(),
                R.animator.card_flip_right_in);

        handler = new UpdateCardsHandler();

        setContentView(R.layout.activity_memory_game);

        gridviewFront = (GridView) findViewById(R.id.gridviewFront);
        gridviewBack = (GridView) findViewById(R.id.gridviewBack);

        buttonBack = (ImageView) findViewById(R.id.buttonBack);

        relativeLayout = (RelativeLayout) findViewById(R.id.relativeLayout);



        level = getIntent().getIntExtra("level", 0);
        Log.i("level", Integer.toString(level));

        newGame();



        gridviewFront.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            public void onItemClick(AdapterView<?> parent, View v,
                                    int position, long id) {

                if(!isTimerStarted){
                    timer = new Chronometer(getApplicationContext());
                    timer.start();
                    isTimerStarted = true;
                }

                if(isClickable){
                    synchronized (lock) {

                        if(firstCard == null){
                            setRightOutFirst.setTarget(v);
                            setRightInFirst.setTarget(gridviewBack.getChildAt(position));
                            setRightOutFirst.start();
                            setRightInFirst.start();
                        }else{

                            if(firstCard.position == position){
                                return; //the user pressed the same card
                            }

                            setRightOutSecond.setTarget(v);
                            setRightInSecond.setTarget(gridviewBack.getChildAt(position));
                            setRightOutSecond.start();
                            setRightInSecond.start();
                        }
                        Log.i("position", Integer.toString(position));
                        turnCard(v,position);
                    }
                }

            }
        });
    }

    @Override
    public void onBackPressed(){}

    private void newGame() {

        memory = new Memory(this);

        // On initialise un tableau d'entiers contenant en position 0 le nombre de colonnes que fera la gridview,
        // et en position 1 le nombre de cartes dans le jeu
        memory.setLevelParams(level);
        gridviewBack.setNumColumns(memory.columns);
        gridviewFront.setNumColumns(memory.columns);
        finish = memory.size / 2;

        List<Integer[]> idDrawablesFrontAndBack = memory.getListDrawablesFrontAndBack();

        // à la position 0 sont placés tous les dos de cartes, à la 1, les différents légumes chargés
        gridviewFront.setAdapter(new ImageAdapter(this, idDrawablesFrontAndBack.get(1)));
        gridviewBack.setAdapter(new ImageAdapter(this, idDrawablesFrontAndBack.get(0)));


    }

    private void turnCard(View v, int position) {



        if(firstCard==null){
            firstCard = new Card(v, gridviewBack.getChildAt(position), position);
        }
        else{

            isClickable = false;

            secondCard = new Card(v, gridviewBack.getChildAt(position), position);


            TimerTask tt = new TimerTask() {

                @Override
                public void run() {
                    try{
                        synchronized (lock) {
                            handler.sendEmptyMessage(0);
                        }
                    }
                    catch (Exception e) {
                        Log.e("E1", e.getMessage());
                    }
                }
            };

            Timer t = new Timer(false);
            t.schedule(tt, 1300);

        }


    }

    @SuppressLint("HandlerLeak")
    private class UpdateCardsHandler extends Handler {

        @Override
        public void handleMessage(Message msg) {
            synchronized (lock) {
                checkCards();
            }
        }
        void checkCards(){
            tries++;

            boolean isSameFruit = memory.checkCards(firstCard, secondCard);

            if(isSameFruit){
                Animation animFadeOut = AnimationUtils.loadAnimation(getApplicationContext(), R.anim.fade_out);
                new ParticleSystem(MemoryGame.this, 1000, getResources().getIdentifier(memory.getNom() + "_ico", "drawable", getPackageName()), 1000)
                        .setSpeedRange(0.2f, 0.5f)
                        .oneShot(firstCard.viewBack, 100);
                firstCard.viewBack.startAnimation(animFadeOut);
                firstCard.viewFront.setVisibility(View.INVISIBLE);
                firstCard.viewBack.setVisibility(View.INVISIBLE);
                new ParticleSystem(MemoryGame.this, 1000, getResources().getIdentifier(memory.getNom() + "_ico", "drawable", getPackageName()), 1000)
                        .setSpeedRange(0.2f, 0.5f)
                        .oneShot(secondCard.viewBack, 100);
                secondCard.viewBack.startAnimation(animFadeOut);
                secondCard.viewFront.setVisibility(View.INVISIBLE);
                secondCard.viewBack.setVisibility(View.INVISIBLE);

                finish--;
                isClickable = true;
            }else{
                setRightOutFirst.setTarget(firstCard.viewBack);
                setRightInFirst.setTarget(firstCard.viewFront);

                setRightOutSecond.setTarget(secondCard.viewBack);
                setRightInSecond.setTarget(secondCard.viewFront);


                setRightOutFirst.start();
                setRightInFirst.start();
                setRightOutSecond.start();
                setRightInSecond.start();

                final Handler handlerFlip = new Handler();
                handlerFlip.postDelayed(new Runnable() {
                    @Override
                    public void run() {
                        isClickable = true;
                    }
                }, 1000);
            }

            firstCard = null;
            secondCard = null;

            if(finish == 0){
                timer.stop();
                result();
            }
        }
    }

    private void result(){

        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(this);

        calendar = Calendar.getInstance();
        String monthString;
        int calendarMonth = calendar.get(Calendar.MONTH) + 1;
        if(calendarMonth < 10){
            monthString = "0" + Integer.toString(calendarMonth);
        }else{
            monthString = Integer.toString(calendarMonth);
        }
        String date = calendar.get(Calendar.DAY_OF_MONTH) + "/" + monthString + "/" + calendar.get(Calendar.YEAR);
        long time = (SystemClock.elapsedRealtime() - timer.getBase()) / 1000;
        Log.i("nb coups", Integer.toString(tries));
        Log.i("timer : ", Long.toString(time));

        Result result = new Result();
        result.setName(preferences.getString(getString(R.string.name), null));
        result.setGame("Memory");
        result.setLevel(level);
        result.setTries(tries);
        result.setTime(time);
        result.setDate(date);

        ResultDAO resultDAO = new ResultDAO(this);
        resultDAO.add(result);


        Bundle bundle = new Bundle();
        bundle.putInt("level", level);
        fragmentResult = new ResultFragment();
        fragmentResult.setArguments(bundle);
        fragmentManager = getSupportFragmentManager();
        buttonBack.setVisibility(View.INVISIBLE);
        final Handler handlerResult = new Handler();
        handlerResult.postDelayed(new Runnable() {
            @Override
            public void run() {
                fragmentManager.beginTransaction()
                        .setCustomAnimations(R.anim.up_start, R.anim.up_end)
                        .replace(R.id.fragment_container, fragmentResult).commit();
            }
        }, 1000);
    }

    public void replayLevel(View view) {

        fragmentManager.beginTransaction().remove(fragmentResult).commit();
        buttonBack.setVisibility(View.VISIBLE);
        newGame();

    }

    public void nextLevel(View view) {

        fragmentManager.beginTransaction().remove(fragmentResult).commit();
        buttonBack.setVisibility(View.VISIBLE);
        isTimerStarted = false;
        tries = 0;
        level++;
        newGame();

    }

    public void back(View v) {

        new AlertDialog.Builder(this).setTitle("Quitter")
                .setMessage("Êtes vous sûrs de vouloir quitter ce jeu ?")
                .setIcon(android.R.drawable.ic_menu_help)
                .setPositiveButton("Oui", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        finish();
                        overridePendingTransition(R.anim.left_start, R.anim.left_end);
                    }
                })
                .setNegativeButton("Non", null).show();

    }
}

File: app/src/main/java/kagura/project/com/fruitlegumes/association/memorielle/MemoryGame.kt
package kagura.project.com.fruitlegumes.association.memorielle

import android.animation.AnimatorInflater
import android.animation.AnimatorSet
import android.annotation.SuppressLint
import android.os.Bundle
import android.os.Handler
import android.os.Message
import android.os.SystemClock
import android.util.Log
import android.view.View
import android.view.Window
import android.view.WindowManager
import android.view.animation.AnimationUtils
import android.widget.AdapterView.OnItemClickListener
import android.widget.Chronometer
import android.widget.GridView
import android.widget.ImageView
import android.widget.RelativeLayout
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.preference.PreferenceManager
import com.plattysoft.leonids.ParticleSystem
import kagura.project.com.fruitlegumes.R
import kagura.project.com.fruitlegumes.adapters.ImageAdapter
import kagura.project.com.fruitlegumes.association.Association
import kagura.project.com.fruitlegumes.association.ResultFragment
import kagura.project.com.fruitlegumes.collections.Card
import kagura.project.com.fruitlegumes.collections.Result
import kagura.project.com.fruitlegumes.database.ResultDAO
import java.util.Calendar
import java.util.Timer
import java.util.TimerTask

class MemoryGame : AppCompatActivity() {
    private var isClickable = true
    private var finish = 0
    private var tries = 0
    private var handler: UpdateCardsHandler? = null
    private var firstCard: Card? = null
    private var secondCard: Card? = null
    private lateinit var setRightOutFirst: AnimatorSet
    private lateinit var setRightInFirst: AnimatorSet
    private lateinit var setRightOutSecond: AnimatorSet
    private lateinit var setRightInSecond: AnimatorSet
    private lateinit var gridviewFront: GridView
    private lateinit var gridviewBack: GridView
    private lateinit var fragmentResult: Fragment
    var buttonBack: ImageView? = null
    private var relativeLayout: RelativeLayout? = null
    private var isTimerStarted = false
    var timer: Chronometer? = null
    private lateinit var calendar: Calendar
    private var fragmentManager: FragmentManager? = null
    lateinit var memory: Association

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        requestWindowFeature(Window.FEATURE_NO_TITLE)
        window.setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN,
                WindowManager.LayoutParams.FLAG_FULLSCREEN)
        window.requestFeature(Window.FEATURE_ACTION_BAR)
        supportActionBar!!.hide()
        handler = UpdateCardsHandler()
        setContentView(R.layout.activity_memory_game)

        setRightOutFirst = AnimatorInflater.loadAnimator(applicationContext,
                R.animator.card_flip_right_out) as AnimatorSet
        setRightInFirst = AnimatorInflater.loadAnimator(applicationContext,
                R.animator.card_flip_right_in) as AnimatorSet
        setRightOutSecond = AnimatorInflater.loadAnimator(applicationContext,
                R.animator.card_flip_right_out) as AnimatorSet
        setRightInSecond = AnimatorInflater.loadAnimator(applicationContext,
                R.animator.card_flip_right_in) as AnimatorSet
        gridviewFront = findViewById<View>(R.id.gridviewFront) as GridView
        gridviewBack = findViewById<View>(R.id.gridviewBack) as GridView
        buttonBack = findViewById<View>(R.id.buttonBack) as ImageView
        relativeLayout = findViewById<View>(R.id.relativeLayout) as RelativeLayout
        level = intent.getIntExtra("level", 0)
        Log.i("level", level.toString())
        newGame()
        Log.i("after", "new game")
        gridviewFront.onItemClickListener = OnItemClickListener { _, v, position, _ ->
            if (!isTimerStarted) {
                timer = Chronometer(applicationContext)
                timer!!.start()
                isTimerStarted = true
            }
            Log.i("after", "chrono")
            if (isClickable) {
                synchronized(lock) {
                    if (firstCard == null) {
                        setRightOutFirst.setTarget(v)
                        setRightInFirst.setTarget(gridviewBack.getChildAt(position))
                        setRightOutFirst.start()
                        setRightInFirst.start()
                    } else {
                        if (firstCard!!.position == position) {
                            return@OnItemClickListener  //the user pressed the same card
                        }
                        setRightOutSecond.setTarget(v)
                        setRightInSecond.setTarget(gridviewBack.getChildAt(position))
                        setRightOutSecond.start()
                        setRightInSecond.start()
                    }
                    Log.i("position", position.toString())
                    turnCard(v, position)
                }
            }
        }
    }

    override fun onBackPressed() {}

    private fun newGame() {
        memory = Memory(this)

        // On initialise un tableau d'entiers contenant en position 0 le nombre de colonnes que fera la gridview,
        // et en position 1 le nombre de cartes dans le jeu
        memory.setLevelParams(level)
        gridviewBack.numColumns = memory.columns
        gridviewFront.numColumns = memory.columns
        finish = memory.size / 2
        val idDrawablesFrontAndBack = memory.listDrawablesFrontAndBack

        // à la position 0 sont placés tous les dos de cartes, à la 1, les différents légumes chargés
        gridviewFront.adapter = ImageAdapter(this, idDrawablesFrontAndBack[1].toTypedArray(), relativeLayout!!.tag.toString())
        gridviewBack.adapter = ImageAdapter(this, idDrawablesFrontAndBack[0].toTypedArray(), relativeLayout!!.tag.toString())
    }

    private fun turnCard(v: View, position: Int) {
        Log.i("debut", "turn card")
        if (firstCard == null) {
            firstCard = Card(v, gridviewBack.getChildAt(position), position)
        } else {
            isClickable = false
            secondCard = Card(v, gridviewBack.getChildAt(position), position)
            val tt: TimerTask = object : TimerTask() {
                override fun run() {
                    try {
                        synchronized(lock) { handler!!.sendEmptyMessage(0) }
                    } catch (e: Exception) {
                        Log.e("E1", e.message.toString())
                    }
                }
            }
            val t = Timer(false)
            t.schedule(tt, 1300)
        }
    }

    @SuppressLint("HandlerLeak")
    private inner class UpdateCardsHandler : Handler() {
        override fun handleMessage(msg: Message) {
            synchronized(lock) { checkCards() }
        }

        fun checkCards() {
            tries++
            val isSameFruit = memory.checkCards(firstCard, secondCard)
            if (isSameFruit) {
                val animFadeOut = AnimationUtils.loadAnimation(applicationContext, R.anim.fade_out)
                ParticleSystem(this@MemoryGame, 1000, resources.getIdentifier(memory.getNom(firstCard!!.position) + "_ico", "drawable", packageName), 1000)
                        .setSpeedRange(0.2f, 0.5f)
                        .oneShot(firstCard!!.viewBack, 100)
                firstCard!!.viewBack.startAnimation(animFadeOut)
                firstCard!!.viewFront.visibility = View.INVISIBLE
                firstCard!!.viewBack.visibility = View.INVISIBLE
                ParticleSystem(this@MemoryGame, 1000, resources.getIdentifier(memory.getNom(secondCard!!.position) + "_ico", "drawable", packageName), 1000)
                        .setSpeedRange(0.2f, 0.5f)
                        .oneShot(secondCard!!.viewBack, 100)
                secondCard!!.viewBack.startAnimation(animFadeOut)
                secondCard!!.viewFront.visibility = View.INVISIBLE
                secondCard!!.viewBack.visibility = View.INVISIBLE
                finish--
                isClickable = true
            } else {
                setRightOutFirst.setTarget(firstCard!!.viewBack)
                setRightInFirst.setTarget(firstCard!!.viewFront)
                setRightOutSecond.setTarget(secondCard!!.viewBack)
                setRightInSecond.setTarget(secondCard!!.viewFront)
                setRightOutFirst.start()
                setRightInFirst.start()
                setRightOutSecond.start()
                setRightInSecond.start()
                val handlerFlip = Handler()
                handlerFlip.postDelayed({ isClickable = true }, 1000)
            }
            firstCard = null
            secondCard = null
            if (finish == 0) {
                timer!!.stop()
                result()
            }
        }
    }

    private fun result() {
        val preferences = PreferenceManager.getDefaultSharedPreferences(this)
        calendar = Calendar.getInstance()
        val monthString: String
        val calendarMonth = calendar.get(Calendar.MONTH) + 1
        monthString = if (calendarMonth < 10) {
            "0$calendarMonth"
        } else {
            calendarMonth.toString()
        }
        val date = calendar.get(Calendar.DAY_OF_MONTH).toString() + "/" + monthString + "/" + calendar.get(Calendar.YEAR)
        val time = (SystemClock.elapsedRealtime() - timer!!.base) / 1000
        Log.i("nb coups", tries.toString())
        Log.i("timer : ", time.toString())
        val result = Result()
        result.name = preferences.getString(getString(R.string.name), null)
        result.game = "Memory"
        result.level = level
        result.tries = tries
        result.time = time
        result.date = date
        val resultDAO = ResultDAO(this)
        resultDAO.add(result)
        val bundle = Bundle()
        bundle.putInt("level", level)
        fragmentResult = ResultFragment()
        fragmentResult.arguments = bundle
        fragmentManager = supportFragmentManager
        buttonBack!!.visibility = View.INVISIBLE
        val handlerResult = Handler()
        handlerResult.postDelayed({
            fragmentManager!!.beginTransaction()
                    .setCustomAnimations(R.anim.up_start, R.anim.up_end)
                    .replace(R.id.fragment_container, fragmentResult).commit()
        }, 1000)
    }

    fun replayLevel(@Suppress("UNUSED_PARAMETER") view: View?) {
        fragmentManager!!.beginTransaction().remove(fragmentResult).commit()
        buttonBack!!.visibility = View.VISIBLE
        newGame()
    }

    fun nextLevel(@Suppress("UNUSED_PARAMETER") view: View?) {
        fragmentManager!!.beginTransaction().remove(fragmentResult).commit()
        buttonBack!!.visibility = View.VISIBLE
        isTimerStarted = false
        tries = 0
        level++
        newGame()
    }

    fun back(@Suppress("UNUSED_PARAMETER") v: View?) {
        AlertDialog.Builder(this).setTitle("Quitter")
                .setMessage("Êtes vous sûrs de vouloir quitter ce jeu ?")
                .setIcon(android.R.drawable.ic_menu_help)
                .setPositiveButton("Oui") { _, _ ->
                    finish()
                    overridePendingTransition(R.anim.left_start, R.anim.left_end)
                }
                .setNegativeButton("Non", null).show()
    }

    companion object {
        private var level = 0
        private val lock = Any()
    }
}